File: minimum-time-to-make-rope-colorful.ts
{
    const minCost = function(colors, neededTime) {
        let total = 0,
            currentMax = 0
    
        for(let i = 0; i < colors.length; i++) {

            if(i > 0 && colors[i] != colors[i - 1]) {
                currentMax = 0
            }

            total += Math.min(currentMax, neededTime[i])
            currentMax = Math.max(currentMax, neededTime[i])
        }

        return total
    };

    console.log(minCost("aaabbbabbbb", [3,5,10,7,5,3,5,5,4,8,1]))
}

File: minimum-time-to-make-rope-colorful.rs
pub fn min_cost(colors: String, needed_time: Vec<i32>) -> i32 {
    let (mut result, mut p, mut m, mut s) = (0, 0, 0, 0);
    for (c, t) in colors.into_bytes().into_iter().zip(needed_time) {
        if c == p {
            m = m.max(t);
            s += t;
        } else {
            result += s - m;
            m = t;
            s = t;
            p = c;
        }
    }
    result + s - m
}
